text/0000-forwardpropstoheader.md

```diff
@@ -0,0 +1,186 @@
+# Reactive Header Control with `<this.props.navigation.ForwardPropsToHeader>`
+## Problem
+In the React lifecycle, all rendering should be done inside `render()`. But if a screen needs to control how a header is rendered, it cannot do so in the `render()` method. Instead, the Screen must use a third-party object to pass callbacks through. This does not scale well for complex headers.
```

Comment:
I think we should be more clear about the problem definition

> In the React lifecycle, all rendering should be done inside `render()`. But if a screen needs to control how a header is rendered, it cannot do so in the `render()` method.

This is a barrier to learning and a bit of an awkward API to not play into the typical React flow. Not in itself motivating enough to prioritize over other changes yet, in my opinion.

> Instead, the Screen must use a third-party object to pass callbacks through.

The typical pattern used right now is `setParams` to pass some callbacks up to the header. Either that or use something like Redux, which I think is similar to what you describe. I think if this is the core of the problem we could come up with different solutions. For example, react-native-navigation and ex-navigation both have a concept of the header and screen components sharing an event emitter. They can communicate by subscribing to / emitting events.

> This does not scale well for complex headers

It would be really great to have an example of one of these complex headers so we have something to compare ideas against!

Reply:
> The typical pattern used right now is `setParams` to pass some callbacks up to the header. Either that or use something like Redux, which I think is similar to what you describe.

I have been using something like Redux, but it's been frustraiting to need to write 10 lines of boilerplate to pass any prop, ref, or callback. JS requires using setters and getters for imported modules; you can't set attributes directly.

> It would be really great to have an example of one of these complex headers so we have something to compare ideas against!

I will see if I can release some source code from the project that I'm working on.